src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -247,32 +236,25 @@ private Task ProcessAuthentication(bool isAsync = false, bool isApm = false, Can
             {
                 ThrowIfExceptional();
 
-                //  A trick to discover and avoid cached sessions.
-                _CachedSession = CachedSessionStatus.Unknown;
-
                 if (isAsync)
                 {
-                    result = ForceAuthenticationAsync(_context.IsServer, null, cancellationToken);
+                    result = ForceAuthenticationAsync(new SslAsyncIO(this, cancellationToken), _context.IsServer, null);
+                    result.ContinueWith(t => { _nestedAuth = 0; }, cancellationToken,
```

Comment:
Why is this continuation needed?  We can't set _nestedAuth to 0 in the ForCeAuthenticationAsync implementation as we were previously doing?

I do like that this approach centralizes the responsibility for zeroing it out here. However, it adds cost, and as written I'm also not 100% convinced it's correct.  We're going to end up setting _nestedAuth to 0 potentially twice, when there's a race condition where this result task completes unsuccessfully by the time we hit the finally block below.

But worse, technically I think this is introducing a race condition where consumers could end up erroneously getting an error that an operation was still in progress even if one wasn't.  The problem is that _nestedAuth is being set to 0 _after_ the result task completes, and it's this result task that's being returned to the caller.  So technically a caller could see that the task completed and go to start their next operation, only to find that _nestedAuth hadn't yet been zeroed.

For this to be correct, we really need to make it so that the returned task doesn't complete until after _nestedAuth is reset.

Reply:
I think I miss something here @stephentoub, and I saw your comment on the previous PR. 
I assumed that either the task will finish synchronously (success or error) and then the finally() will clean the `_nestedAuth`. Or it won't and then (and only then) the continuation will kick in.

I do not understand the comment about the previous set of _nestedAuth in ForceAuthenticationAsync().
Originally in 3.x, we would set it back either in `ProcessAuthentication` as we do now or in `EndProcessAuthentication`. But it was never cleared inside of the `ForceAuthenticationAsync`.

On a side note, I do envision more cleanup needed when `ForceAuthenticationAsync` is done. (both success and failure). Perhaps the cost is small comparing to all the crypto. And simplification of the logic would be worth it. I can run some benchmarks to see if that matters in practice.

Reply:
> I assumed that either the task will finish synchronously (success or error) and then the finally() will clean the _nestedAuth. Or it won't and then (and only then) the continuation will kick in.

The finally block is checking IsCompleted.  If the result task completes asynchronously prior to that finally block getting executed, IsCompleted will be true, so the finally block will set that field.  And if it completed faulted/canceled, the continuation will also run.

> Originally in 3.x, we would set it back either in ProcessAuthentication as we do now or in EndProcessAuthentication

The EndProcessAuthentication is what I was alluding to.  That was called as part of AuthenticateAsServer/ClientAsync, and before the returned Task was completed, because it was part of the EndXx call used in the Begin/EndXx pair.

Reply:
I moved try/finally  from `ProcessAuthentication()` to `ForceAuthenticationAsync` so the cost should be identical. Since all tasks are awaited there, cleanup should be much simplified and I think it will become handy for next step.